c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -330,10 +440,19 @@ impl<'a, 'tcx> HirTyVisitor<'a, 'tcx> {
             self.hir_rewrites.push((hir_ty.span, rw));
         }
 
+        if let TyKind::Adt(adt_def, substs) = rw_lty.ty.kind() {
+            if let Some(adt_metadata) = &self.acx.gacx.adt_metadata.table.get(&adt_def.did()) {
+                self.hir_rewrites.push((
+                    hir_ty.span,
+                    adt_ty_rw(adt_def, &adt_metadata.lifetime_params, substs),
```

Comment:
`adt_metadata.lifetime_params` tells you what lifetimes are declared on the type, not what lifetimes should be passed at this particular use of it.  Due to the way we infer hypothetical lifetimes in ADTs, it happens to be the case that the names match up within struct fields (given `struct Foo<'h0> { ... }`, all uses will look like `struct Bar<'h0, 'h1> { foo: &'h1 Foo<'h0> }`, with the same name `'h0` in both places), but this isn't true in general.

```Rust
struct Foo<'h0> { ... }
fn foo1<'a>(x: Foo<'a>, y: Foo<'a>) { ... }
fn foo2<'a, 'b>(x: Foo<'a>, y: Foo<'b>) { ... }
```

Given `Foo`, both the `foo1` and `foo2` signatures are valid, and they mean different things, so we need a way to express both of them.

`RewriteLabel` probably needs a field like `&'tcx [OriginArg<'tcx>]` representing the lifetime arguments of a particular `Ty`, with similar meaning to the labels of `FieldMetadata`'s `origin_args` field.  That will allow for distinctions like the above `foo1`/`foo2` example.

For struct field types, the lifetime arguments can be set from the existing `FieldMetadata::origin_args` labels.  For types in function signatures, we'll eventually need a similar setup to generate hypothetical parameters and so on, but for now that doesn't exist, so any region arguments should either be elided or left as `'static`.  Inside function bodies, the regions can always be elided.

Reply:
Thanks for the catch, the tests were wrong and didn't complain about this. I've addressed it in 37e14bc1 and took out the test for the function lifetime rewriting for the time being; I'll create an issue for it

Reply:
Issue created #921